src/java.net.http/share/classes/java/net/http/HttpRequest.java

```diff
@@ -720,6 +720,42 @@ public static BodyPublisher ofFile(Path path) throws FileNotFoundException {
             return RequestPublishers.FilePublisher.create(path);
         }
 
+        /**
+         * {@return a request body publisher whose body is the {@code length}
+         * content bytes read from the provided file {@code channel} starting
+         * from the specified {@code offset}}
+         * <p>
+         * The {@linkplain FileChannel file channel} will be read using
+         * {@link FileChannel#read(ByteBuffer, long) FileChannel.read(ByteBuffer buffer, long position)},
+         * which does not modify the channel's position. Thus, the same file
+         * channel may be shared between several publishers passed to
+         * concurrent requests.
+         * <p>
+         * The file channel will not be closed upon completion. The caller is
+         * expected to manage the life cycle of the channel, and close it
+         * appropriately when not needed anymore.
+         *
+         * @param channel a file channel
+         * @param offset the offset of the first byte
+         * @param length the number of bytes to read from the file channel
+         *
+         * @throws IndexOutOfBoundsException if the specified byte range is
+         * found to be {@linkplain Objects#checkFromIndexSize(long, long, long)
+         * out of bounds} compared with the size of the file referred by the
+         * channel
+         *
+         * @throws IOException if the size of the file referred by the provided
+         * channel cannot be read while verifying the specified byte range
+         *
+         * @throws NullPointerException if {@code channel} is null
+         *
+         * @since 26
+         */
+        public static BodyPublisher ofFileChannel(FileChannel channel, long offset, long length) throws IOException {
+            Objects.requireNonNull(channel, "channel");
```

Comment:
```java
        public FileChannelPublisher(FileChannel channel, long offset, long length) throws IOException {
            this.channel = Objects.requireNonNull(channel, "channel");
            // ...
```
The FileChannelPublisher constructor already has a null value check for the parameter channel, which is redundant

Reply:
We should raise an exception at the API boundary, hence this check. Why didn't I validate `offset` and `length` here too? Because doing so would expose implementation details in the interface. Why did I add an extra null check in `FCP::new`? That's also an API boundary, but an internal one. Naturally, these practices can vary depending on performance considerations.